File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
function solution(maps) {
    let n = maps.length;
    let m = maps[0].length;
    let visited = Array.from(Array(n), () => new Array(m).fill(false));
    let q = [];

    let dx = [1, -1, 0, 0];
    let dy = [0, 0, 1, -1];

    function bfs() {
        visited[0][0] = true;
        q.push([0, 0, 1]);

        while (q.length > 0) {
            let [cx, cy, dis] = q.shift(); // BFS는 shift

            if (cx === n - 1 && cy === m - 1) {
                return dis;
            }

            for (let i = 0; i < 4; i++) {
                let nx = cx + dx[i];
                let ny = cy + dy[i];

                if (nx >= 0 && nx < n && ny >= 0 && ny < m) {
                    if (!visited[nx][ny] && maps[nx][ny] === 1) {
                        visited[nx][ny] = true;
                        q.push([nx, ny, dis + 1]);
                    }
                }
            }
        }

        return -1;
    }

    return bfs();
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    
    x_len = len(maps)
    y_len = len(maps[0])
    
    visited = [[0] * y_len for _ in range(x_len)]
    
    dx = [1 , -1 , 0 , 0]
    dy = [0 , 0 , -1 ,1]
    
    def bfs(x , y):
        q = deque()
        q.append((x, y, 1))
        visited[x][y] = 1
        
        while q:
            cx , cy , ct = q.popleft()
            
            if cx == x_len - 1 and cy == y_len - 1:
                return ct
            
            for i in range(4):
                nx = cx + dx[i]
                ny = cy + dy[i]
                
                if 0 <= nx < x_len and 0 <= ny < y_len:
                    if visited[nx][ny] == 0 and maps[nx][ny] == 1:
                        q.append((nx, ny, ct + 1))
                        visited[nx][ny] = 1
        return -1
    
    return bfs(0, 0)